File: src/test/java/io/github/petrknap/persistence/zoneddatetime/UtcTestCase.java
package io.github.petrknap.persistence.zoneddatetime;

import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.Test;

import java.time.ZonedDateTime;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertSame;

abstract class UtcTestCase extends TestCase
{
    abstract @Test void constructs_itself();

    abstract @Test void fromValues();

    abstract @Test void fromValues_of_null();

    abstract @Test void fromFormattedValues();

    abstract @Test void fromFormattedValues_of_null();

    @Test void asNullable_returns_this()
    {
        Utc instance = getInstance(zonedDateTime);

        assertSame(instance, instance.asNullable());
    }

    @Test void getUtcDateTime_as_formatted_string()
    {
        assertEquals(
                localDateTimeFormatter.format(utcDateTime),
                getInstance(zonedDateTime).getUtcDateTime(LOCAL_DATETIME_FORMAT)
        );
    }

    @Test void toZonedDateTime()
    {
        assertEquals(
                zonedDateTime,
                getInstance(zonedDateTime).toZonedDateTime()
        );
    }

    abstract protected @NotNull Utc getInstance(@NotNull ZonedDateTime zonedDateTime);
}

File: src/test/php/UtcTestCase.php
<?php

declare(strict_types=1);

namespace PetrKnap\Persistence\ZonedDateTime;

use DateTimeImmutable;

/**
 * @phpstan-import-type ZonedDateTime from JavaSe8\Time
 */
abstract class UtcTestCase extends TestCase
{
    abstract public function test_constructs_itself(): void;

    abstract public function test_fromValues(): void;

    abstract public function test_fromValues_of_null(): void;

    abstract public function test_fromFormattedValues(): void;

    abstract public function test_fromFormattedValues_of_null(): void;

    public function test_asNullable_returns_this(): void
    {
        $instance = $this->getInstance($this->zonedDateTime);

        self::assertSame($instance, $instance->asNullable());
    }

    public function test_getUtcDateTime_as_formatted_string(): void
    {
        self::assertEquals(
            $this->utcDateTime->format(self::LOCAL_DATETIME_FORMAT),
            $this->getInstance($this->zonedDateTime)->getUtcDateTime(self::LOCAL_DATETIME_FORMAT)
        );
    }

    public function test_toZonedDateTime(): void
    {
        self::assertDateTimeEquals(
            $this->zonedDateTime,
            $this->getInstance($this->zonedDateTime)->toZonedDateTime(),
        );
    }

    /**
     * @param ZonedDateTime $zonedDateTime
     */
    abstract protected function getInstance(DateTimeImmutable $zonedDateTime): Utc;
}
